### ISSUE TITLE
Player position changed when bumping into borders of tilemap

### ISSUE THREAD

**Issue by tylerroost** [Annoraaq/phaser-grid-movement-plugin#26]
Hello,
thank you for such a wonderfully easy plugin to use. It does everything I need and more
the only problem I have found is that when my player walks to the edge of the tilemap and collides with the border. The position of the player is updated and thus unaligns with the tilemap.
I'm sure there is a solution, but I am not familiar enough with phaser or your plugin to figure it out, my apologies.
![PlayerOutline](https://user-images.githubusercontent.com/31251196/105640457-0b3fb680-5e33-11eb-85cb-2be86b157abc.PNG)
![PlayerInline](https://user-images.githubusercontent.com/31251196/105640459-0bd84d00-5e33-11eb-8ed2-483c8c707a65.PNG)

In the top picture the player can be seen between two tiles

**Comment by Annoraaq** [comment]
Hey @tylerroost,

thank you for the positive feedback!
I am glad that you find it useful and it is always nice to get feedback from people who use it in their projects. You are a great help to make this plugin better. 

Let me take a look at the issue.

**Comment by Annoraaq** [comment]
@tylerroost
I can't reproduce the issue. Can you provide more information to help me tracking it down?

i.e:

- Version of the plugin
- code that you used to initialize the plugin, including initialization of the sprites.

Thank you 🙂

**Comment by tylerroost** [comment]
It may be from the fact that my tileset is 32x32, where as my player sprite is 64x64, thus there maybe half(32) tiles that allow the character to reposition.

**Comment by tylerroost** [comment]
If this seems like a reasonable explanation let me know, and if you still want code I would be happy to provide

**Comment by tylerroost** [comment]
Plugin version: 1.2.2
```
var config = {
    type: Phaser.CANVAS,
    plugins: {
      scene: [
        {
            key: "gridMovementPlugin",
            plugin: GridMovementPlugin,
            mapping: "gridMovementPlugin",
        },
        {
            key: 'rexUI',
            plugin: rexuiplugin,
            mapping: 'rexUI'
        }
      ],
    },
    scale: {
      width: 720,
      height: 528,
      autoCenter: Phaser.Scale.CENTER_BOTH,
    },
    parent: 'content',
    pixelArt: true,
    physics: {
        default: 'arcade',
        arcade: {
            gravity: { y: 0 }
        }
    },
    scene: [
        BootScene,
        WorldScene,
        BattleScene,
        BattleUIScene
    ],
    scale: {
        mode: Phaser.Scale.FIT,
        autoCenter: Phaser.Scale.CENTER_BOTH,
    },
};
```

Here you can see the config for my Phaser.Game object, where the plugin is being invoked.

```
class WorldScene extends Phaser.Scene{
    constructor()
    {
        super('WorldScene')
    }

    WorldScene()
    {
        Phaser.Scene.call(this, { key: 'WorldScene' });
    }
    preload()
    {
        this.load.tilemapTiledJSON('WorldMap', 'assets/locations/WorldMap5.json');
        this.load.spritesheet('HGSSTileset',
            'assets/locations/HGSSTileset.png',
            { frameWidth: 32, frameHeight: 32 }
        );
        this.load.spritesheet('boy',
            'assets/Overworld/PlayerSprites/boy_run.png',
            { frameWidth: 64, frameHeight: 64 }
        );
        this.load.spritesheet('girl',
            'assets/Overworld/PlayerSprites/girl_run.png',
            { frameWidth: 64, frameHeight: 64 }
        );
        this.load.spritesheet('gen1SpriteMap',
            'assets/Pokemon/SpriteMaps/gen1SpriteMap.png'),
            { frameWidth: 128, frameHeight: 128 }
    }
    create()
    {
        var map = this.make.tilemap({ key: 'WorldMap' });
        var tiles = map.addTilesetImage('HGSSTileset');

        var ground = map.createLayer('GroundLayer', tiles, 0, 0);
        var grass = map.createLayer('GrassLayer', tiles, 0, 0);
        var buildings = map.createLayer('BuildingsLayer', tiles, 0, 0);
        ground.scale = 1;
        ground.setDepth(0)
        grass.scale = 1;
        grass.setDepth(1)
        buildings.scale = 1;
        grass.setDepth(2)
        buildings.setCollisionByExclusion();
        this.player = this.physics.add.sprite(400, 350, 'girl', 1);
        this.player.scale = 1;
        this.physics.world.bounds.width = map.widthInPixels;
        this.physics.world.bounds.height = map.heightInPixels;
        this.player.setCollideWorldBounds(true);
        this.cameras.main.startFollow(this.player);
        const gridMovementConfig = {
            characters: [
                {
                    id: "player",
                    sprite: this.player,
                    characterIndex: 0,
                    speed: 4,
                    startPosition: new Phaser.Math.Vector2(17, 17),
                },
            ],
        firstLayerAboveChar: 3,
        };

        this.gridMovementPlugin.create(map, gridMovementConfig);
    }
    update(time, delta)
    {
        const cursors = this.input.keyboard.createCursorKeys();
        if (cursors.left.isDown) {
            this.gridMovementPlugin.moveLeft("player");
        } else if (cursors.right.isDown) {
            this.gridMovementPlugin.moveRight("player");
        } else if (cursors.up.isDown) {
            this.gridMovementPlugin.moveUp("player");
        } else if (cursors.down.isDown) {
            this.gridMovementPlugin.moveDown("player");
        }
    }
};
```
Here is the scene that was used in the imagery. I have attached all necessary files
[GridMovementBug.zip](https://github.com/Annoraaq/phaser-grid-movement-plugin/files/5869240/GridMovementBug.zip)


If this is insufficient to duplicate please let me know what else you need.

**Comment by tylerroost** [comment]
I very much appreciate it, though if I were to guess it has to do with the tilemap being nondivisible by the size of the player character. If this is in fact the cause I'm guessing a simple check against the divisible boundaries of the tilemap could lead to a solution.

**Comment by Annoraaq** [comment]
I assumed the same first. But it had nothing to do with it. Only the tile size of the map is important. However, I think I found the issue:

The problem is this part of your code: `this.player.setCollideWorldBounds(true);`

It enables the collision of the arcade physic engine. So this means the sprite will collide with the boundaries of your map on a per-pixel level. I recommend to just delete this line, because what you might want to have in a tile-based movement are tile-based collisions.

phaser-grid-movement-plugin does support collisions on a tile level. Your tiles just need a custom boolean property called `collides` with the value `true`. In case you are using Tiled you can simply add that property to your tileset. I described it in my article here: https://medium.com/swlh/grid-based-movement-in-a-top-down-2d-rpg-with-phaser-3-e3a3486eb2fd

Moreover, boundaries of the whole map are detected by the plugin by default (they are not tiles so they don't need a `collides` property.

Please let me know if that solved your problem 🙂

**Comment by tylerroost** [comment]
O god how could I have missed that. I'll look into it, and get back to you.

**Comment by Annoraaq** [comment]
@tylerroost can you already give me any feedback whether the issue is resolved?

**Comment by tylerroost** [comment]
My apologies, Ive been swamped with work for school and have an interview on Friday, I should have some free time this weekend. I will schedule in time on Saturday. But based on your analysis, I would say your conclusion is fairly accurate. Like you said the plugin detects collision with the world boundary by default. 

I just did some further research and I don't know if this is relevant, but it appears that when adding a body to a container it may reset one of these two properties 
```
this.player.body.setCollideWorldBounds(true);
this.player.body.onWorldBounds = true;
```
Though it seems like I am colliding with the world boundary, so I'm not entirely sure what is going on, my best guess, without looking at the code for the plugin is that because I never set onWorldBounds to true, the plugin is not being notified that I am at the worldBounds. The other possibility is that even if I did set this property, if I added the player to a container it is implicitly resetting the property, which means the plugin is not being notified of the collision. 

Though like you said the plugin is programmed in such a way that collision with the world bounds is enabled by default, and setting setCollideWorldBounds to true is detecting pixel level collisions and not tile level collisions like I want. I would argue the intended result of setting setCollideWorldBounds to true, while using your plugin is such that the plugin should override the behavior and thus be able to handle the collision. I'm not sure how you would go about this, but let me lay out an example to illustrate my point. Say I have many body gameObects in a container, without a deeper understanding of the plugin, I and possibly others would expect that setting setCollideWorldBounds to true would produce the intended tile based behavior.

TL;DR
There could be multiple reasons for why I am colliding at a pixel level with the boundary.
My intended behavior is more or less the typical expectation and thus should be factored into design for the plugin. In my mind this is easily accomplishable by setting the property (setCollideWorldBounds) to false. Though I am sure there are arguments against this. My apologies, if this comes off as presumptuous.

**Comment by Annoraaq** [comment]
No problem. Don't rush.

Thank you for the detailed answer. I agree with you that it may be misleading for many users that the setCollideWorldBounds property has not the desired effect when using the plugin.

I think I will try to turn it off in the plugin by default (or find a different way of ignoring it). If some users need to turn it back on (for whatever reason) I can still add a configuration setting for this purpose.

I keep this issue up to date.

**Comment by tylerroost** [comment]
Sounds good, thank you for considering my request.

**Comment by Annoraaq** [comment]
I looked a bit into it and thought more about it. I am not sure anymore if this is the correct way to go for the following reasons:

The property only exists on arcade physics objects. I think in most cases mixing up arcade physics with grid physics is a bad idea. Even if I check for this property and disable it, all the other behavior that comes with arcade physics would still be in place (for instance the collision with other arcade physics objects).

The plugin works when just using a plain sprite (Phaser.GameObjects.Sprite). Because the arcade sprite (Phaser.Physics.Arcade.Sprite) extends Phaser.GameObjects.Sprite, you can use it with the plugin. However, that does not mean you should. There might be good reasons to use an arcade sprite (maybe you already have one in your project?).
But when you mix in another physics engine then the borderCollision is not the only thing I would have to deal with.

Therefore I better don't mess around with the physics settings in the plugin, because if someone uses an arcade sprite there might be a specific reason for this.

The problem that you have with turning the setting back on is a separate issue I think.  To avoid that I would have to repeatedly turn it off, wich feels hacky and might also be confusing for a user who wants to enable it on purpose (for whatever reason).

In your specific case, would it be an option to use a plain Phaser.GameObjects.Sprite for your player or do you need the arcade physics?

**Comment by tylerroost** [comment]
Hmm, it does seem quite hacky as you described it, I need to do more research to determine the nature of my sprite. I will add it to my list of things to do this weekend.

**Comment by Annoraaq** [comment]
I hope it's ok if I close this issue for now? I think it is not a real issue with the plugin but rather a clash of settings when using arcade physics with it. If you find out how we could make it more compatible feel free to reopen this issue or create a new one with the suggestion.
